### PR TITLE
At file Usefulness

### PR DESCRIPTION
**Pull request by valaises**

### CODE REVIEW

File: src/scratchpads/chat_utils_rag.rs [hunk 1]
```diff
@@ -111,7 +158,7 @@ pub async fn postprocess_rag_stage1(
         for (line_n, line) in fref.markup.file_content.lines().enumerate() {
             let a = Arc::new(FileLine {
                 fref: fref.clone(),
-                line_n: line_n,
```
Comment (olegklimov) [inline src/scratchpads/chat_utils_rag.rs:114]: No

File: src/scratchpads/chat_utils_rag.rs [hunk 2]
```diff
@@ -66,6 +66,53 @@ pub fn context_to_fim_debug_page(t0: &Instant, postprocessed_messages: &[Context
...
+            _ => 0.0,
+        };
+        set_useful_for_line(line, usefulness);
+    }
+}
+
+fn set_useful_for_line(line: &Arc<FileLine>, useful: f32) {
+    let lineref_mut: *mut FileLine = Arc::as_ptr(line) as *mut FileLine;
+    unsafe { (*lineref_mut).useful = useful; }
+}
```
Comment (olegklimov) [inline src/scratchpads/chat_utils_rag.rs:114]: Should only set if more useful

File: src/scratchpads/chat_utils_rag.rs [hunk 3]
```diff
@@ -66,6 +66,53 @@ pub fn context_to_fim_debug_page(t0: &Instant, postprocessed_messages: &[Context
...
+            3 => if line_n < omsg.line1 {-1.} else {(line_n as f32 * m12 + c12).max(5.)},
+            4 => {
+                if line_n < omsg.line1 {
+                    line_n as f32 * m11 + c11
+                } else if line_n >= omsg.line1 && line_n <= omsg.line2 {
+                    100.
+                } else {
+                    line_n as f32 * m22 + c22
+                }
+            }.max(5.),
```
Comment (olegklimov) [inline src/scratchpads/chat_utils_rag.rs:104]: max(5) is not a good idea

File: src/scratchpads/chat_utils_rag.rs [hunk 4]
```diff
@@ -66,6 +66,53 @@ pub fn context_to_fim_debug_page(t0: &Instant, postprocessed_messages: &[Context
...
+        let m = (y2 - y1) / (x2 - x1);
+        let c = y1 - m * x1;
+        (m, c)
+    }
+
+    if omsg.gradient_type < 0 || omsg.gradient_type > 4 {
+        return;
+    }
+
+    let t_fade_away_lines = 50;
```
Comment (olegklimov) [inline src/scratchpads/chat_utils_rag.rs:83]: t_ ?

Comment (valaises) [reply src/scratchpads/chat_utils_rag.rs:83]: threshold

**Comment by olegklimov** [comment]
merged

**Pull request closed without merging**